src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java

```diff
@@ -0,0 +1,330 @@
+        headingController.setGoal(desiredAngle.getRadians());
+        double output = headingController.calculate(Robot.getAngle().getRadians());
+
+        setStates(kinematics.toSwerveModuleStates(new ChassisSpeeds(forward, strafe, output)));
+    }
+
+    /**
+     * Gets te states of every module.
+     *
+     * @return the states of every module.
+     */
+    public SwerveModuleState[] getStates() {
+        SwerveModuleState[] swerveModuleState = new SwerveModuleState[modules.length];
+        for (SwerveModule module : modules) {
+            swerveModuleState[module.getWheel()] = module.getState();
+        }
+        return swerveModuleState;
+    }
+
+    /**
+     * Sets the state of the modules.
+     *
+     * @param states the states of the modules.
+     */
+    public void setStates(SwerveModuleState[] states) {
+        SwerveDriveKinematics.desaturateWheelSpeeds(states, 1.42 * Constants.SwerveDrive.VELOCITY_MULTIPLIER);
+        for (SwerveModule module : modules) {
+            SwerveModuleState state = states[module.getWheel()];
+            state = SwerveModuleState.optimize(state, module.getAngle());
+            module.setState(state);
+        }
+    }
+
+    /**
+     * Gets the chassis speeds of the entire robot.
+     *
+     * @return the speed of the robot in each axis.
+     */
+    public ChassisSpeeds getChassisSpeeds() {
+        ChassisSpeeds chassisSpeeds = kinematics.toChassisSpeeds(getStates());
+        if (fieldOriented) {
+            chassisSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(
+                    chassisSpeeds.vxMetersPerSecond,
+                    chassisSpeeds.vyMetersPerSecond,
+                    chassisSpeeds.omegaRadiansPerSecond,
+                    Robot.getAngle()
+            );
+        }
+        return chassisSpeeds;
+    }
+
+    /**
+     * Gets a specific module, shouldn't be used for regular cases.
+     *
+     * @param index the index of the module.
+     * @return the corresponding module.
+     */
+    public SwerveModule getModule(int index) {
+        return modules[index];
+    }
+
+    /**
+     * Gets the pose of the robot.
+     *
+     * @return the pose of the robot.
+     */
+    public Pose2d getPose() {
+        return odometry.getPoseMeters();
+    }
+
+    /**
+     * Resets the odometry.
+     */
+    public void resetOdometry() {
+        resetOdometry(new Pose2d(), new Rotation2d());
+    }
+
+    /**
+     * Resets the odometry to a specified position.
+     *
+     * @param pose     the current pose.
+     * @param rotation the holonomic rotation.
+     */
+    public void resetOdometry(Pose2d pose, Rotation2d rotation) {
+        odometry.resetPosition(new Pose2d(pose.getTranslation(), rotation), rotation);
+    }
+
+    /**
+     * Resets the heading controller target angle.
+     */
+    public void resetHeadingController() {
+        headingController.reset(0, getChassisSpeeds().omegaRadiansPerSecond);
```

Comment:
I don't think it will work, the heading controller controls angle, and `omegaRadiansPerSecond` controls the rotational velocity of the robot.